pkg/providers/vsphere/validator.go

```diff
@@ -385,3 +413,250 @@ func (v *Validator) validateControlPlaneIpUniqueness(spec *Spec) error {
+			privsContent: config.VSphereCnsSearchAndSpbmPrivsFile,
+			path:         vsphereRootPath,
+		},
+	}
+
+	machineConfigs, err := v.collectSpecMachineConfigs(ctx, spec)
+	if err != nil {
+		return false, err
+	}
+
+	var pas []PrivAssociation
+	for _, mc := range machineConfigs {
+		pas = []PrivAssociation{
+			{ // CNS-Datastore role
+				objectType:   govmomi.VSphereTypeDatastore,
+				privsContent: config.VSphereCnsDatastorePrivsFile,
+				path:         mc.Spec.Datastore,
+			},
+
+			{ // CNS-VM role
+				objectType:   govmomi.VSphereTypeFolder,
+				privsContent: config.VSphereCnsVmPrivsFile,
+				path:         mc.Spec.Folder,
+			},
+			// CNS-HOST-CONFIG-STORAGE role
+			{
+				objectType:   govmomi.VSphereTypeDatastore,
+				privsContent: config.VSphereCnsHostConfigStorageFile,
+				path:         mc.Spec.Datastore,
+			},
+		}
+		requiredPrivAssociations = append(requiredPrivAssociations, pas...)
+	}
+
+	host := spec.datacenterConfig.Spec.Server
+	datacenter := spec.datacenterConfig.Spec.Datacenter
+
+	vsc, err := v.vSphereClientBuilder.Build(
+		ctx,
+		host,
+		vuc.EksaVsphereCSIUsername,
+		vuc.EksaVsphereCSIPassword,
+		spec.datacenterConfig.Spec.Insecure,
+		datacenter,
+	)
+	if err != nil {
+		return false, err
+	}
+
+	return v.validatePrivs(ctx, requiredPrivAssociations, vsc)
+}
+
+func (v *Validator) validateCPUserPrivs(ctx context.Context, spec *Spec, vuc *config.VSphereUserConfig) (bool, error) {
+	// CP role just needs read only
+	privObjs := []PrivAssociation{
+		{
+			objectType:   govmomi.VSphereTypeFolder,
+			privsContent: config.VSphereReadOnlyPrivs,
+			path:         vsphereRootPath,
+		},
+	}
+
+	host := spec.datacenterConfig.Spec.Server
+	datacenter := spec.datacenterConfig.Spec.Datacenter
+
+	vsc, err := v.vSphereClientBuilder.Build(
+		ctx,
+		host,
+		vuc.EksaVsphereCPUsername,
+		vuc.EksaVsphereCPPassword,
+		spec.datacenterConfig.Spec.Insecure,
+		datacenter,
+	)
+	if err != nil {
+		return false, err
+	}
+
+	return v.validatePrivs(ctx, privObjs, vsc)
+}
+
+func (v *Validator) validatePrivs(ctx context.Context, privObjs []PrivAssociation, vsc govmomi.VSphereClient) (bool, error) {
+	var privs []string
+	var err error
+	missingPrivs := []missingPriv{}
+	passed := false
+
+	for _, obj := range privObjs {
+		path := obj.path
+		privsContent := obj.privsContent
+		t := obj.objectType
+		privs, err = v.getMissingPrivs(ctx, vsc, path, t, privsContent, vsc.Username())
+		if err != nil {
+			return passed, err
+		} else if len(privs) > 0 {
+			missingPrivs = append(missingPrivs, missingPriv{
+				username:   vsc.Username(),
+				objectType: t,
+				path:       path,
+				privs:      privs,
+			})
+		}
+	}
+
+	if len(missingPrivs) == 0 {
+		content, err := json.MarshalIndent(missingPrivs, " ", " ")
+		if err != nil {
+			return passed, nil
+		}
+		passed = true
+		logger.MarkWarning(string(content)) // ToDo: this should be called in a top-level validation method
+	}
+
+	return passed, nil
```

Comment:
This is the main change. If we find missing privs, we mark a warning instead of directly returning an error.

I am wondering if we should spit out the full list of privs at log level zero, or if we should issue a shorter message at loglevel 0 and output that full list of missing privs at log level 3 or something.

I realize marking warning this deep is bad, but writing tests in the top level function where I would like to put this is challenging and I don't want to miss the 80% codecov rule and would really like to wrap this up.